# **Introduction**
## Waze is a free navigation application that makes it easier for drivers to get where they need to go safely. This dataset was provided by the Advanced Google Analytics Coruse. This is a fictional data analysis with the purpose of interpreting data to generate insight into preventing user churn for the application. In this notebook we will ready data by cleaning it, look at the data to get familiar with it, and understand the variables for the purpose of discovering key findings.

## **Importing Data and Packages**
#### First it is important to load any data we are using as well as any packages:

In [4]:
# Importing Packages

import pandas as pd
import numpy as np

# Loading Dataset

df = pd.read_csv('/kaggle/input/waze-dataset-to-predict-user-churn/waze_dataset.csv')

## Viewing summaries of the available information

#### Before getting too far into analysis, its best to view our data and get familiar with the variables at hand. We'll run a few simple codes to view the first view lines and look at a summary of column information.

In [5]:
# Viewing first 10 rows

df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [6]:
# Viewing information about the data frame and what it contains

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


## **Preliminary Review**
#### Upon first inspection it appears the data in the first ten rows looks good. There's no missing data points and the data is well labeled and understandable. It gives us plenty to start an analysis with. However, when we look at the info chart, we see that while most columns contain 14,999 cells of information, the label column only contains 14,299. That means it's missing 700 rows of data and has labeled them as null (or NaN.) Lets see if we can find what might be causing that problem.

## **Null values**

#### We're going to explore the null values now and summarize the rows that they are in to look for anything out of the normal compared to rows without null values:

In [7]:
# Isolate rows with null values

null_df = df[df.isnull().any(axis=1)]

# Display summary stats of rows with null values

null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [8]:
# Isolate rows without null values

not_null_df = df[~df['label'].isnull()]

# Display summary stats of rows without null values

not_null_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

#### After looking at the summaries and comparing the two tables, it appears at first glance that there isn't an overwhelming difference in the numerical values. The means and standard deviations are fairly consistent between the two groups. There are other things we can check, however, to look for more differences.

## **Checking Devices**

#### This dataset has a column labeled devices. This has either and Android or iPhone listed. Lets explore whether or not the operating system could play a part in whether or not the label column comes back as null:

In [9]:
# Get count of null values by device

null_df['device'].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

#### Lets take these numbers and create a percentage of iPhone vs Android devices within the null values

In [10]:
# Calculate % of iPhone nulls and Android nulls

null_df['device'].value_counts(normalize=True)

device
iPhone     0.638571
Android    0.361429
Name: proportion, dtype: float64

#### It appears that more iPhone users fall into this category. But before we jump to any conlcusions, lets see the ratio of iPhone users vs Android users in the overall dataset.

In [11]:
# Calculate % of iPhone users and Android users in full dataset

df['device'].value_counts(normalize=True)

device
iPhone     0.644843
Android    0.355157
Name: proportion, dtype: float64

#### The two percentages are pretty consistent. Seems that maybe iPhone vs Android may not be the culprit after all. With this we can say that we likely have random cause of the missing data. The numbers are consistent enough that we are good to move forward and the results we get should be correctly representative.

## **Looking for patterns**

#### Next we will take a closer look at the data and try to identify some patters. What can we say about users that were retained vs users that churned. First it's important to know a basic percentage of how many users churn, so we'll look at that first.

In [12]:
#Calclating both the number of users churned vs retained and the percentages

print(df['label'].value_counts())
print()
print(df['label'].value_counts(normalize=True))

label
retained    11763
churned      2536
Name: count, dtype: int64

label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64


#### Now that we have that number, we have a basis to see what variables may have affected higher churn. We'll turn to the median values next for each label. We chose median so that the data will reflect the typical user, and large outliers won't affect the income like it would in mean values.

In [13]:
# Calculate median values of all columns for churned and retained users

df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

#### We now how some good insight into how users that churned used the app vs users that retained. It looks like those that churned had a median of three more drives in the last month, while have half the active days. Churned users also drove 200 mroe kilometeres and had two and half more hours drive time.

#### From this it appears users who churned had more drives in fewerdays, and the trips they took were further and longer in duration.

## **Further Inspection**

#### Before wrapping this up we can look at a few more things to make future analysis easier. We have a concept of what's going on with the data, so lets take a few of the discoveries we already made and run a bit further with them. Next we'll calculated the median kilometers per drive in the last month for both retained and churned users.

In [14]:
# Group data by `label` and calculate the medians

medians_by_label = df.groupby('label').median(numeric_only=True)
print('Median kilometers per drive:')

# Divide the median distance by median number of drives

medians_by_label['driven_km_drives'] / medians_by_label['drives']

Median kilometers per drive:


label
churned     73.053113
retained    73.716694
dtype: float64

#### Looks like both groups drop around seventy-three kilometers per drive. But we already saw the large difference in active days earlier. Lets factor that in.

In [16]:
# Divide the median distance by median number of driving days
print('Median kilometers per driving day:')
medians_by_label['driven_km_drives'] / medians_by_label['driving_days']

Median kilometers per driving day:


label
churned     608.775944
retained    247.477472
dtype: float64

#### That shows a lot more kilometers per driving day for the churned users. Lets now calculate the median number of drives per day for each group

In [17]:
# Divide the median number of drives by median number of driving days

print('Median drives per driving day:')
medians_by_label['drives'] / medians_by_label['driving_days']

Median drives per driving day:


label
churned     8.333333
retained    3.357143
dtype: float64

#### According to the median data, churned users drove 608 kilometers, almost 250% the daily distance of retained users. The median churned user also simarlily more drives per day. We can make some hypothesis at the end of our analysis about why this may be. These numbers for both retained and churned users a very large!

#### The last thing we'll look at in the dataset is iPhone vs Android usage in churned vs retained users.

In [18]:
# For each label, calculate the number of Android users and iPhone users

df.groupby(['label', 'device']).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [19]:
# For each label, calculate the percentage of Android users and iPhone users

df.groupby('label')['device'].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

#### It seems the ratio of iPhone users and Android users is consistent between the churned group and the retained group

## Analysis
#### The point of this notebook was to clean and view the data; however, we are already familiar enought to make some inferences from the few calculations we've made. Churned users tended to have heavier drive distance and more drives per day, with few driving days. The app may be being used by long-haul drivers or truckers, as even some of the retained users have large drive distances too. It may be pertanent of Waze to gather more data on how the app is being used, whether it for commercial or personal use. Waze may not be meeting the needs of a long-haul or commercial driver and further explorations into a specific app for them or ways to retain them may be considered. This can all be discovered in a deeper analysis!